### PR TITLE
Deprecate LightningDistributed and keep logic in ddp/ddpSpawn directly

### LINKED ISSUE #9692: Deprecate LightningDistributed and keep broadcast in ddp/ddpSpawn directly
## Proposed refactoring or deprecation
1. 
LightningDistributed() class only used by ddp and ddpSpawn, and only have one broadcast function for torch collectives. It's unnecessary to have.
 Also, we have to set rank and device in set up steps. If subclass extend DDP or DDPSpawn, and overridden function where LightningDistributed.rank and device setted, could cause silent failures.

2. Now the src is not respected in torch broadcast

### Motivation
Simplify the code structure and reduce the possibilities for silent failure

### Pitch
Deprecate LightningDistributed. There is only one function
```
    def broadcast(self, obj: Any, group=_group.WORLD):
        # always wrap into a list so it can be broadcasted.
        obj = [obj]

        if self.rank != 0:
            obj = [None] * len(obj)

        broadcast_object_list(obj, 0, group=group or _group.WORLD)

        return obj[0]

```
Move to ddp and ddpSpawn
```
   def broadcast(self, obj: object, src: int = 0) -> object:
        if not distributed_available():
            raise RuntimeError("DDP is not initialized and torch.distributed is not available, can not broadcast object")
        obj = [obj]
        if self.global_rank != 0:
            obj = [None] * len(obj)
        broadcast_object_list(obj, src, group=_group.WORLD)
        return obj[0]
```

### Additional context

Related to https://github.com/PyTorchLightning/pytorch-lightning/issues/7534

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

### PR DESCRIPTION
**Pull request by four4fish**
## What does this PR do?

LightningDistributed() class only used by ddp and ddpSpawn, and only have one broadcast function for torch collectives. It's unnecessary to have.
Also, we have to set rank and device in set up steps. If subclass extend DDP or DDPSpawn, and overridden function where LightningDistributed.rank and device setted, could cause silent failures.

Now the src is not respected in torch broadcast, fix the issue

Fixes #9692

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/distributed/dist.py [hunk 1]
```diff
@@ -18,6 +18,12 @@
 
 
 class LightningDistributed:
+    """
+    .. deprecated:: v1.5
+        This method is deprecated in v1.5and will be removed in v1.7.
+        Logic moved to ddp and ddpspawning
```
Comment (awaelchli) [inline pytorch_lightning/distributed/dist.py:24]:
```suggestion
        This class is deprecated in v1.5 and will be removed in v1.7.
        The broadcast logic will be moved to the :class:`DDPPlugin` and  :class`DDPSpawnPlugin` classes.
```

Comment (awaelchli) [reply pytorch_lightning/distributed/dist.py:24]: @four4fish we need to print a rank_zero_deprection message here in the init of the class and/or in the broadcast message below

File: CHANGELOG.md [hunk 2]
```diff
@@ -267,6 +267,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Deprecated passing `progress_bar_refresh_rate` to the `Trainer` constructor in favor of adding the `ProgressBar` callback with `refresh_rate` directly to the list of callbacks ([#9616](https://github.com/PyTorchLightning/pytorch-lightning/pull/9616))
 
 
+- Deprecate `LightningDistributed` and move the broadcast logic to ddp/ddpSpawn directly ([#9691](https://github.com/PyTorchLightning/pytorch-lightning/pull/9691))
+
```
Comment (ananthsub) [inline CHANGELOG.md:271]:
```suggestion
- Deprecate `LightningDistributed` and move the broadcast logic to `DDPPlugin` and `DDPSpawnPlugin` directly ([#9691](https://github.com/PyTorchLightning/pytorch-lightning/pull/9691))

```

**Review by awaelchli — APPROVED** [review]
Have a look at the `tests/deprecated_api` where we should add a quick test for the 1.7 removal :)

File: pytorch_lightning/distributed/dist.py [hunk 3]
```diff
@@ -18,6 +18,12 @@
...
+    """
+
     def __init__(self, rank=None, device=None):
```
Comment (ananthsub) [inline pytorch_lightning/distributed/dist.py:27]: 1. add a deprecation message in the init so a warning shows up at runtime
2. Add a test for the deprecation here: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/tests/deprecated_api/test_remove_1-7.py

so we know to delete this class in v1.7

File: pytorch_lightning/plugins/training_type/ddp.py [hunk 4]
```diff
@@ -403,7 +396,11 @@ def barrier(self, *args, **kwargs) -> None:
             torch.distributed.barrier()
 
     def broadcast(self, obj: object, src: int = 0) -> object:
-        return self.dist.broadcast(obj)
+        obj = [obj]
```
Comment (ananthsub) [inline pytorch_lightning/plugins/training_type/ddp.py:399]: - please address the bug with `src` not being respected
- as discussed elsewhere, we can special case Tensor objects for a potential speedup as @yifuwang noted here: https://github.com/PyTorchLightning/pytorch-lightning/pull/9414#discussion_r707788699

Comment (four4fish) [reply pytorch_lightning/plugins/training_type/ddp.py:399]: >please address the bug with src not being respected

It's address here, as src will be pass into `broadcast_object_list`, before the src was 0

>as discussed elsewhere, we can special case Tensor objects for a potential speedup as @yifuwang noted here: 2/n Consolidate collective functions - collective base and subclasses #9414

Yeah, good point. Will create a separate new feature issue address this

File: pytorch_lightning/distributed/dist.py (hunk 3 above)
Comment (four4fish) [reply pytorch_lightning/distributed/dist.py:27]: @awaelchli @ananthsub  This is a internal use only class, do you guys think we need wait until 1.7 to remove this or it is safe to be directly removed?

Comment (awaelchli) [reply pytorch_lightning/distributed/dist.py:27]: yes I think we need to unfortunately. it's been around for a while and is not marked as experimental.

Comment (ananthsub) [reply pytorch_lightning/distributed/dist.py:27]: we can remove its usage from within the plugins, but we cannot delete the class outright. it will have to sit unused in the codebase until v1.7

Comment (awaelchli) [reply pytorch_lightning/distributed/dist.py:27]: @ananthsub's comment reminds me that we should then exclude this file also from coverage. we can add it in `setup.cfg ` and add it to the list of omitted files

File: pytorch_lightning/plugins/training_type/ddp.py [hunk 5]
```diff
@@ -403,7 +396,11 @@ def barrier(self, *args, **kwargs) -> None:
...
+        if self.global_rank != 0:
```
Comment (ananthsub) [inline pytorch_lightning/plugins/training_type/ddp.py:400]:
```suggestion
        if self.global_rank != src:
```
does this need to be the src instead of hardcoded to rank 0?

File: pytorch_lightning/plugins/training_type/ddp.py [hunk 6]
```diff
@@ -403,7 +396,11 @@ def barrier(self, *args, **kwargs) -> None:
...
+            obj = [None] * len(obj)
```
Comment (ananthsub) [inline pytorch_lightning/plugins/training_type/ddp.py:401]: this is not correct: obj is not guaranteed to be a list. and this will cause surprised with strings

Comment (four4fish) [reply pytorch_lightning/plugins/training_type/ddp.py:401]: it been converted to a string at line 399

obj = [obj]

File: pytorch_lightning/plugins/training_type/ddp.py (hunk 5 above)
Comment (four4fish) [reply pytorch_lightning/plugins/training_type/ddp.py:400]: I think the old assumption is that torch distributed broadcast only works with broadcast from global rank 0. WIth respect the change respect src, i think here should be change to src as well. Good catch

File: pytorch_lightning/plugins/training_type/ddp.py [hunk 7]
```diff
@@ -396,14 +389,20 @@ def post_dispatch(self, trainer: "pl.Trainer") -> None:
...
+            raise RuntimeError("DDP is not initialized and torch.distributed is not avalible, can not broadcast object")
         if _TORCH_GREATER_EQUAL_1_8 and torch.distributed.get_backend() == "nccl":
             torch.distributed.barrier(device_ids=self.determine_ddp_device_ids())
         else:
             torch.distributed.barrier()
 
     def broadcast(self, obj: object, src: int = 0) -> object:
-        return self.dist.broadcast(obj)
+        if not distributed_available():
+            raise RuntimeError("DDP is not initialized and torch.distributed is not avalible, can not broadcast object")
```
Comment (ananthsub) [inline pytorch_lightning/plugins/training_type/ddp.py]: this should throw automatically as part of  torch distributed's broadcast implementation

File: pytorch_lightning/plugins/training_type/ddp.py (hunk 6 above)
Comment (ananthsub) [reply pytorch_lightning/plugins/training_type/ddp.py:401]: it's possible we can save an extra list serialization by checking if `isinstance(obj, list)` and calling `broadcast_object_list` directly in that case w/o needing to convert it

File: pytorch_lightning/plugins/training_type/ddp_spawn.py [hunk 8]
```diff
@@ -324,7 +315,11 @@ def barrier(self, *args, **kwargs) -> None:
     def broadcast(self, obj: object, src: int = 0) -> object:
         if not distributed_available():
             return obj
-        return self.dist.broadcast(obj)
+        obj = [obj]
+        if self.global_rank != 0:
```
Comment (ananthsub) [inline pytorch_lightning/plugins/training_type/ddp_spawn.py:319]: same here

File: pytorch_lightning/plugins/training_type/ddp.py (hunk 6 above)
Comment (four4fish) [reply pytorch_lightning/plugins/training_type/ddp.py:401]: Current logic seems works really wired with list obj. I'm not sure whether check list before convert is a breaking change or not
```
if obj = [1,2,3]
obj= [obj] -> obj = [[1,2,3]]
```

File: pytorch_lightning/distributed/dist.py [hunk 9]
```diff
@@ -14,11 +14,22 @@
...
+    """
+    .. deprecated:: v1.5
+        This class is deprecated in v1.5 and will be removed in v1.7.
+        The broadcast logic will be moved to the :class:`DDPPlugin` and :class`DDPSpawnPlugin` classes.
+    """
+
     def __init__(self, rank=None, device=None):
+        rank_zero_deprecation(
+            "This class is deprecated in v1.5 and will be removed in v1.7."
+            "he broadcast logic will be moved to the :class:`DDPPlugin` and :class`DDPSpawnPlugin` classes."
```
Comment (ananthsub) [inline pytorch_lightning/distributed/dist.py:31]:
```suggestion
            "Broadcast logic is implemented directly in the :class:`TrainingTypePlugin` implementations."
```
?

**Review by ananthsub — APPROVED** [review]
i think there's an optimization opportunity for broadcast but we can pick it up in a follow up PR along with tensor specializations

File: pytorch_lightning/plugins/training_type/ddp.py (hunk 6 above)
Comment (ananthsub) [reply pytorch_lightning/plugins/training_type/ddp.py:401]: why would it be a breaking change? since we can post-process the return of `broadcast_object_list` before we return, there should be no leakage to the caller. its a minor efficiency point is all. but we can discuss this further with better support for broadcasting tensors in a separate PR

File: CHANGELOG.md [hunk 10]
```diff
@@ -267,6 +267,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Deprecated passing `progress_bar_refresh_rate` to the `Trainer` constructor in favor of adding the `ProgressBar` callback with `refresh_rate` directly to the list of callbacks ([#9616](https://github.com/PyTorchLightning/pytorch-lightning/pull/9616))
 
 
+- Deprecate `LightningDistributed` and move the broadcast logic to `DDPPlugin` and `DDPSpawnPlugin` directly ([#9691](https://github.com/PyTorchLightning/pytorch-lightning/pull/9691))
```
Comment (ananthsub) [inline CHANGELOG.md:270]: also add a changelog entry that broadcast in DDP/DDP spawn plugins now respect the `src` input

File: CHANGELOG.md [hunk 11]
```diff
@@ -392,6 +395,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Fixed wrapping issue: avoid wrapping LightningModule with data-parallel modules when not fitting in `DDPPlugin`, `DDPSpawnPlugin`, `DDPShardedPlugin`, `DDPSpawnShardedPlugin` ([#9096](https://github.com/PyTorchLightning/pytorch-lightning/pull/9096))
 
 
+- Fixed `broadcast` in `DDPPlugin` and ``DDPSpawnPlugin` to respects the src input  ([#9691](https://github.com/PyTorchLightning/pytorch-lightning/pull/9691))
```
Comment (ananthsub) [inline CHANGELOG.md:398]:
```suggestion
- Fixed `broadcast` in `DDPPlugin` and ``DDPSpawnPlugin` to respect the `src` input  ([#9691](https://github.com/PyTorchLightning/pytorch-lightning/pull/9691))
```

File: pytorch_lightning/plugins/training_type/ddp.py (hunk 6 above)
Comment (ananthsub) [reply pytorch_lightning/plugins/training_type/ddp.py:401]: nit: won't len(obj) always be 1?

Comment (four4fish) [reply pytorch_lightning/plugins/training_type/ddp.py:401]: It's not nit...I feel we can revisit this part of the logic in a separate issue/PR. Do we need set a None list on the ranks broadcast too?

File: pytorch_lightning/plugins/training_type/ddp.py [hunk 12]
```diff
@@ -270,8 +265,6 @@ def setup_distributed(self):
         init_ddp_connection(self.cluster_environment, self.torch_distributed_backend)
 
         # set the ranks and devices
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/ddp.py:267]: leftover comment

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/plugins/training_type/ddp.py (hunk 6 above)
Comment (ananthsub) [reply pytorch_lightning/plugins/training_type/ddp.py:401]: > Do we need set a None list on the ranks broadcast too?

yes, each rank must provide lists of equal sizes: https://pytorch.org/docs/stable/_modules/torch/distributed/distributed_c10d.html#broadcast_object_list

but we don't need to multiply by len(obj) here.

File: pytorch_lightning/distributed/dist.py [hunk 13]
```diff
@@ -14,11 +14,22 @@
...
 class LightningDistributed:
+    """
+    .. deprecated:: v1.5
+        This class is deprecated in v1.5 and will be removed in v1.7.
+        The broadcast logic will be moved to the :class:`DDPPlugin` and :class`DDPSpawnPlugin` classes.
+    """
+
     def __init__(self, rank=None, device=None):
+        rank_zero_deprecation(
+            "This class is deprecated in v1.5 and will be removed in v1.7."
```
Comment (ananthsub) [inline pytorch_lightning/distributed/dist.py:30]:
```suggestion
            "LightningDistributed is deprecated in v1.5 and will be removed in v1.7."
```

File: tests/deprecated_api/test_remove_1-7.py [hunk 14]
```diff
@@ -243,3 +243,10 @@ def test_v1_7_0_lightning_logger_base_close(tmpdir):
     ):
         logger = LoggerCollection([logger])
         logger.close()
+
+
+def test_v1_7_0_deprecate_lightning_distributed(tmpdir):
+    with pytest.deprecated_call(match="This class is deprecated in v1.5 and will be removed in v1.7."):
```
Comment (ananthsub) [inline tests/deprecated_api/test_remove_1-7.py:249]:
```suggestion
    with pytest.deprecated_call(match="LightningDistributed is deprecated in v1.5 and will be removed in v1.7."):
```

File: pytorch_lightning/plugins/training_type/ddp.py (hunk 6 above)
Comment (ananthsub) [reply pytorch_lightning/plugins/training_type/ddp.py:401]:
```suggestion
            obj = [None]
```

File: pytorch_lightning/plugins/training_type/ddp_spawn.py [hunk 15]
```diff
@@ -324,7 +315,11 @@ def barrier(self, *args, **kwargs) -> None:
     def broadcast(self, obj: object, src: int = 0) -> object:
         if not distributed_available():
             return obj
-        return self.dist.broadcast(obj)
+        obj = [obj]
+        if self.global_rank != src:
+            obj = [None] * len(obj)
```
Comment (ananthsub) [inline pytorch_lightning/plugins/training_type/ddp_spawn.py:320]:
```suggestion
            obj = [None]
```
